mfkdf2/src/derive/key.rs

```diff
@@ -1,77 +1,504 @@
-  for factor in policy.clone().factors {
-    let mut material = match factors.get(factor.id.as_str()).cloned() {
+  let mut outputs = HashMap::new();
+  let mut factors = factors.clone(); // TODO (@lonerapier): fix this
+  let mut new_policy = policy.clone();
+
+  for factor in new_policy.factors.iter_mut() {
+    let material = match factors.get_mut(factor.id.as_str()) {
       Some(material) => material,
       None => continue,
     };
 
-    material.factor_type.include_params(serde_json::from_str(&factor.params).unwrap());
+    if material.kind() == "persisted" {
+      shares_bytes.push(material.data());
+    } else {
+      material.factor_type.include_params(serde_json::from_str(&factor.params).unwrap())?;
+
+      // TODO (autoparallel): This should probably be done with a `MaybeUninit` array.
+      let salt_bytes = general_purpose::STANDARD.decode(&factor.salt)?;
+
+      let stretched = hkdf_sha256_with_info(
+        &material.data(),
+        &salt_bytes,
+        format!("mfkdf2:factor:pad:{}", factor.id).as_bytes(),
+      );
 
-    // TODO (autoparallel): This should probably be done with a `MaybeUninit` array.
-    let salt_bytes = general_purpose::STANDARD.decode(&factor.salt)?;
-    let salt_arr: [u8; 32] = salt_bytes.try_into().map_err(|_| MFKDF2Error::TryFromVecError)?;
+      // TODO (autoparallel): This should probably be done with a `MaybeUninit` array.
+      let pad = general_purpose::STANDARD.decode(&factor.pad)?;
+      // TODO (@lonerapier): unpadding of bytes is needed
+      let plaintext = decrypt(pad, &stretched);
 
-    let stretched = hkdf_sha256(&material.factor_type.bytes(), &salt_arr);
+      if !factor.hint.is_empty() {
+        let buffer = hkdf_sha256_with_info(
+          &stretched,
+          &material.salt,
+          format!("mfkdf2:factor:hint:{}", factor.id).as_bytes(),
+        );
 
-    // TODO (autoparallel): This should probably be done with a `MaybeUninit` array.
-    let pad = general_purpose::STANDARD.decode(&factor.pad)?;
-    let plaintext = decrypt(pad, &stretched);
+        let binary_string: String =
+          buffer.iter().map(|byte| format!("{:08b}", byte)).collect::<Vec<_>>().join("");
 
-    // TODO (autoparallel): It would be preferred to know the size of this array at compile
-    // time.
-    shares_bytes.push(plaintext);
+        // Take the last `hint_len` characters
+        let hint = binary_string
+          .chars()
+          .rev()
+          .take(factor.hint.len())
+          .collect::<Vec<_>>()
+          .into_iter()
+          .rev()
+          .collect::<String>();
+
+        if hint != factor.hint {
+          return Err(MFKDF2Error::HintMismatch(factor.id.clone()));
+        }
+      }
+
+      // TODO (autoparallel): It would be preferred to know the size of this array at compile
+      // time.
+      shares_bytes.push(plaintext);
+      outputs.insert(factor.id.clone(), material.factor_type.output_derive().to_string());
```

Comment:
This is where you can benefit from dropping the unified `Factor` trait, if you care and think it's cleaner. What I mean is you could have this:

```suggestion
      outputs.insert(factor.id.clone(), <material.factor_type as FactorDerive>::derive().to_string());
```

My syntax may be slightly off and it doesn't matter much either way, but yeah.